toppra/dracula/__init__.py

```diff
@@ -146,6 +147,35 @@ def lims_obeyed(self, traj, raise_2nd_order):
                 return False
         return True
 
+    def joint_limits_obeyed(self, traj, multiplier):
+        lims_obeyed = True
+        t_max = 1 if multiplier is None else multiplier * self.t_sum
+        t_sample = np.arange(0, t_max, 0.001)
+        q_sample = traj(t_sample)
+
+        t_index = 0
+        waypt_copy = [self.waypts[0]]
+        for ii in range(1, self.waypts.shape[0]):
+            t_waypt = -1
+            closest_dist = 1e6
+            for t_test in range(t_index, len(q_sample)):
+                q_dist = np.linalg.norm(q_sample[t_test] - self.waypts[ii])
+                if q_dist < closest_dist:
+                    closest_dist = q_dist
+                    t_waypt = t_test
+
+            i, j = np.where(~(
+                (self.qlim[:, 0] < q_sample[t_index:t_waypt + 1]) &
+                (q_sample[t_index:t_waypt + 1] < self.qlim[:, 1])))
+            if i.size or j.size:
+                # t_index + i are violating knot points
+                lims_obeyed = False
+                waypt_copy += [(self.waypts[ii-1] + self.waypts[ii])/2.]
+            t_index = t_waypt
+            waypt_copy += [self.waypts[ii]]
```

Comment:
`.append()` is the pythonic way or writing it but sill inefficient compared to modifying ndarray slices